inventory-service/src/main/java/com/petclinic/inventoryservice/businesslayer/ProductInventoryServiceImpl.java

```diff
@@ -499,5 +499,38 @@ public Flux<ProductResponseDTO> getLowStockProducts(String inventoryId, int stoc
                 .switchIfEmpty(Mono.error(new NotFoundException("No products below threshold in inventory: " + inventoryId)));
     }
 
+    @Override
+    public Flux<ProductResponseDTO>searchProducts(String inventoryId, String productName, String productDescription) {
+        if (productName != null && productDescription != null){
+            return productRepository
+                    .findAllProductsByInventoryIdAndProductNameAndProductDescription(inventoryId, productName, productDescription)
+                    .map(EntityDTOUtil::toProductResponseDTO)
+                    .switchIfEmpty(Mono.error(new NotFoundException("Inventory not found with InventoryId: " + inventoryId +
+                            "\nOr ProductName: " + productName + "\nOr ProductDescription: " + productDescription)));
+        }
+
+        if (productName != null){
+            return productRepository
+                    .findAllProductsByInventoryIdAndProductName(inventoryId, productName)
+                    .map(EntityDTOUtil::toProductResponseDTO)
+                    .switchIfEmpty(Mono.error(new NotFoundException("Inventory not found with InventoryId: " + inventoryId +
+                            "\nOr ProductName: " + productName)));
+        }
+
+        if (productDescription != null){
+            return productRepository
+                    .findAllProductsByInventoryIdAndProductDescription(inventoryId, productDescription)
+                    .map(EntityDTOUtil::toProductResponseDTO)
+                    .switchIfEmpty(Mono.error(new NotFoundException("Inventory not found with InventoryId: " + inventoryId +
+                            "\nOr ProductDescription: " + productDescription)));
+        }
+
+        return productRepository
+                .findAllProductsByInventoryId(inventoryId)
+                .map(EntityDTOUtil::toProductResponseDTO)
+                .switchIfEmpty(Mono.error(new NotFoundException("Inventory not found with InventoryId: " + inventoryId)));
+
+    }
+
```

Comment:
The implementation of your searchProducts looks fine and you also added error handling in the event that either an InventoryId is not found, a product name is not found or a product description is not found. Your implementation handles each one of these scenarios which is great for debugging purposes. I would consider changing your error messages to be a bit more descriptive. Overall good work!